runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -197,104 +327,139 @@ impl API for Module {
     }
 }
 
-impl Module {
-    const EVM_CONFIG: Config = Config::istanbul();
+impl<Cfg: Config> Module<Cfg> {
+    const EVM_CONFIG: EVMConfig = EVMConfig::istanbul();
 
-    fn do_evm<C, F, V>(
-        source: H160,
-        value: U256,
-        gas_price_in: U256,
-        gas_limit: u64,
-        ctx: &mut C,
-        f: F,
-    ) -> Result<V, Error>
+    fn do_evm<C, F, V>(source: H160, value: U256, ctx: &mut C, f: F) -> Result<V, Error>
     where
         F: FnOnce(
             &mut StackExecutor<'static, MemoryStackState<'_, 'static, evm_backend::Backend<'_, C>>>,
+            u64,
         ) -> (evm::ExitReason, V),
-        C: Context,
+        C: TxContext,
     {
-        // Make sure that gas_price >= min_gas_price.
         let params = Self::params(ctx.runtime_state());
-        let gas_price: primitive_types::U256 = gas_price_in.into();
+        let den = params.token_denomination;
         let min_gas_price: primitive_types::U256 = params.min_gas_price.into();
+
+        let gas_limit: u64 = core::Module::remaining_tx_gas(ctx);
+        let gas_price: primitive_types::U256 = ctx.tx_auth_info().fee.gas_price().into();
+
+        // Make sure that gas_price >= min_gas_price.
         if gas_price < min_gas_price {
             return Err(Error::GasPriceTooLow);
         }
 
         let vicinity = evm_backend::Vicinity {
-            gas_price: gas_price_in,
+            gas_price: gas_price.into(),
             origin: source,
         };
 
+        // The maximum gas fee has already been withdrawn in authenticate_tx().
+        let paid_fee = ctx.tx_auth_info().fee.amount.amount();
+        let max_gas_fee = gas_price
+            .checked_mul(primitive_types::U256::from(gas_limit))
+            .ok_or(Error::FeeOverflow)?;
+
+        if paid_fee < max_gas_fee.as_u128() {
```

Comment:
Why is this check needed, briefly it seems like this will never be true as on one hand we know that:
```
paid_fee = ai.fee.amount
max_gas_fee = gas_price * gas_limit
gas_price = ai.fee.amount / ai.fee.gas
gas_limit <= ai.fee.gas
```

And on the other:
```
paid_fee < max_gas_fee
ai.fee.amount < (ai.fee.amount / ai.fee.gas) * gas_limit
1 < gas_limit / ai.fee.gas
ai.fee.gas < gas_limit
```

Reply:
Oops, that's a leftover :)